src/main/kotlin/de/twintorx/battleship/ui/Player.kt

```diff
@@ -58,32 +58,49 @@ class Player {
     }
 
     private fun prepare() {
-        val ships = Ship.getStandardShipSet()
+        var ships = Ship.getStandardShipSet()
 
         // place ships manually
         if (Console.clearConsole().run { Console.input("${PlayerMessage.PLACE_OPTIONS}\n") { InputRegex.YES_OR_NO.matches(it) }.toLowerCase() == "y" }) {
             Console.printPlaceShips(gameBoard, trackBoard, remainingEnemyHitPoints, remainingOwnHitPoints, remainingEnemyShips, remainingOwnShips)
-
+            val shipStack = mutableListOf<Pair<Ship, HashSet<Point>>>() // TODO initialize shippstack at beginning of player
             while (ships.isNotEmpty()) {
                 Console.printChooseShip()
                 val option = Console.input(ships.map {
-                    val name = it.value[0].type
-                    "[${it.key}] ${it.value.size}x$name${" " * (11 - name.length)}(Size:${it.value[0].size})\n"
-                }.joinToString("")) {
-                    InputRegex.SELECT_SHIP.matches(it) && ships.containsKey(it.toInt())
+                    with(it.value[0].type) {
+                        "[${it.key}] ${it.value.size}x$this${" " * (11 - length)}(Size:${it.value[0].size})\n"
+                    }
+                }.joinToString("") + if (shipStack.isNotEmpty()) "${PlayerMessage.UNDO_OPTION}\n" else "") {
+                    InputRegex.SELECT_SHIP.matches(it) && ((ships.containsKey(it.toInt())) || (it.toInt() == 6 && shipStack.isNotEmpty()))
                 }.toInt()
-
                 Console.eraseLastLines(ships.size + 2)
 
-                ships[option]?.let {
-                    placeShip(it[0])
-
-                    Console.printPlaceShips(gameBoard, trackBoard, remainingEnemyHitPoints, remainingOwnHitPoints, remainingEnemyShips, remainingOwnShips)
-
-                    it.removeAt(0)
-
-                    if (it.isEmpty()) {
-                        ships.remove(option)
+                if (option == 6 && shipStack.isNotEmpty()) {
+                    val pair = shipStack[shipStack.size - 1]
```

Comment:
val (... , ...) = shipStack[gugu gaga]
then you dont have meaningless first and second